Native memory management documentation (#28628)
quarkusio/quarkus: docs/src/main/asciidoc/native-reference.adoc

xref:native-and-ssl.adoc[Using SSL With Native Images],
and xref:writing-native-applications-tips.adoc[Writing Native Applications],
guides.
It explores advanced topics that help users diagnose issues,
increase the reliability and improve the runtime performance of native executables.
These are the high level sections to be found in this guide:

* <<native-memory-management,Native Memory Management>>
* <<inspecting-and-debugging,Inspecting and Debugging Native Executables>>
* <<native-faq,Frequently Asked Questions>>

[[native-memory-management]]
== Native Memory Management
Memory management for Quarkus native executables is enabled by GraalVM’s SubstrateVM runtime system.
The memory management component in GraalVM is explained in detail

> link:https://www.graalvm.org/22.2/reference-manual/native-image/optimizations-and-performance/MemoryManagement[here].

Author: Really? What do you have to say about future URL changes to the non-versioned links and content deviation?

> This guide complements the information in the GraalVM website with further observations particularly relevant for Quarkus applications.

Reviewer: ```suggestion
This guide complements the information available in the GraalVM website with further observations particularly relevant to Quarkus applications.
```

=== Garbage Collectors
The garbage collectors available for Quarkus users are currently Serial GC and Epsilon GC.

==== Serial GC

> Serial GC, the default option in GraalVM, is a single-threaded non-concurrent GC, just like HotSpot’s Serial GC.

Reviewer: ```suggestion
Serial GC, the default option in GraalVM and Quarkus, is a single-threaded non-concurrent GC, just like HotSpot’s Serial GC.
```

The implementation in GraalVM however is different to the HotSpot one,
and there can be significant differences in the runtime behavior.

> One of the key differences between HotSpot’s Serial GC and GraalVM’s Serial GC is precisely on how full GCs are executed.

Reviewer: ```suggestion
One of the key differences between HotSpot’s Serial GC and GraalVM’s Serial GC is the way they perform full GC cycles.
```

In HotSpot the algorithm used is mark-sweep-compact whereas in GraalVM it is mark-copy.
Both need to traverse all live objects,
but in mark-copy this traversal is also used to copy live objects to a secondary space or semi-space.
As objects are copied from one semi-space to another they’re also compacted.
In mark-sweep-compact, the compacting requires a second pass on the live objects.
This makes full GCs in mark-copy more time efficient (in terms of time spent in each GC cycle) than mark-sweep-compact.
The tradeoff mark-copy makes in order to make individual full GC cycles shorter is space.
The use of semi-spaces means that for an application to maintain the same GC performance that mark-sweep achieves (in terms of allocated MB per second),
it requires double the amount of memory.

> 
> Just like in GraalVM, Quarkus also uses Serial GC as the default gargabe collector.

Reviewer: ```suggestion
```

> Just like in GraalVM, Quarkus also uses Serial GC as the default gargabe collector.

Author: Dropped.

===== GC Collection Policy

> The default GC collection policy is called “adaptive”,
> and it is based on HotSpot's ParallelGC adaptive size policy.

Reviewer: ```suggestion
===== GC Collection Policy

GraalVM's Serial GC implementation offers a choice between two different collection policies, the default is called "adaptive" and the alternative is called "space/time".

The “adaptive” collection policy is based on HotSpot's ParallelGC adaptive size policy.
```

The main difference with HotSpot is its focus on memory footprint.
This means that GraalVM’s adaptive GC policy tries to aggressively trigger GCs in order to keep memory consumption down.

> Up to version 2.13, Quarkus used “space/time” GC collection policy,
> but starting with version 2.14, Quarkus uses “adaptive” policy.

Reviewer: ```suggestion
Up to version 2.13, Quarkus used the “space/time” GC collection policy by default,
but starting with version 2.14, it switched to using the “adaptive” policy instead.
```

> Quarkus chose to use space/time early on because it had considerable performance improvements over adaptive,
> but recently we have discovered that space/time can have a worse out-of-the-box experience compared with space/time,
> while at the same time the benefits it offered have diminished considerably.

Reviewer: ```suggestion
The reason why Quarkus initially chose to use "space/time" is because at that time it had considerable performance improvements over "adaptive".
Recent performance experiments, however, indicate that the "space/time" policy can result in worse out-of-the-box experience compared to the "space/time" policy,
while at the same time the benefits it used to offer have diminished considerably after improvements made to the "adaptive" policy.
```

> Better out-of-the-box experience and improvements to adaptive over the past few years make it the best option for most, if not all, Quarkus applications.

Reviewer: ```suggestion
As a result, the "adaptive" policy appears to be the best option for most, if not all, Quarkus applications.
```

Full details on this switch can be read in link:https://github.com/quarkusio/quarkus/issues/28267[this issue].

> It is still possible to configure the space/time GC collection policy using GraalVM’s `-H:InitialCollectionPolicy` flag.
> Switching to the space/time can be done by passing the following via command line:

Reviewer: ```suggestion
It is still possible to change the GC collection policy using GraalVM’s `-H:InitialCollectionPolicy` flag.
Switching to the "space/time" policy can be done by passing the following via command line:
```

[source,bash]
----
-Dquarkus.native.additional-build-args=-H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy\$BySpaceAndTime
----

Or adding this to the `application.properties` file:

[source,properties]
----
quarkus.native.additional-build-args=-H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime
----

[NOTE]
====

> Escaping the `$` character is required to configure space/time GC collection policy if passing via command line in Bash.

Reviewer: ```suggestion
Escaping the `$` character is required to configure the "space/time" GC collection policy if passing via command line in Bash.
```

Other command line environments might have similar requirements.
====

==== Epsilon GC
Epsilon GC is a no-op garbage collector which does not do any memory reclamation.
From a Quarkus perspective, some of the most relevant use cases for this garbage collector are extremely short-lived jobs, e.g. serverless functions.
To build Quarkus native with epsilon GC, pass the following argument at build time:

[source,bash]
----
-Dquarkus.native.additional-build-args=--gc=epsilon
----

=== Memory Management Options
Options to control maximum heap size, young space and other typical use cases found in the JVM can be found in

> https://www.graalvm.org/22.2/reference-manual/native-image/optimizations-and-performance/MemoryManagement[the GraalVM memory management guide].

Reviewer: ```suggestion
https://www.graalvm.org/reference-manual/native-image/optimizations-and-performance/MemoryManagement[the GraalVM memory management guide].
```

Dropping the version from the url redirects to the latest version of the docs. This comes with the risk of having a broken link in the future, but makes sure we won't need to update the version every once in a while. Alternatively, we could somehow trim `{graalvm-version}` to only keep the first two parts of the version, e.g., keep `22.2` from `22.2.0`.

Author: There are 4 other cases in the documentation, some of which still point to 22.1.

Unless I'm mistaken, `{graalvm-version}` is derived from `graal-sdk.version-for-documentation`:

```
<graalvm-version>${graal-sdk.version-for-documentation}</graalvm-version>
```

Which in turn comes from:

```
<graal-sdk.version-for-documentation>22.2.0</graal-sdk.version-for-documentation>
```

But the interesting thing is in the line below this one:

```
<mandrel.version-for-documentation>22.2</mandrel.version-for-documentation>
```

That is in turn exported to docs via `mandrel-version`:

```
<mandrel-version>${mandrel.version-for-documentation}</mandrel-version>
```

So, we could "hijack" `{mandrel-version}` for the major.minor combo, or have a dedicated `{graalvm-majorminor-version}`.

WDYT? @gsmet?

Author: In any case, I don't like the idea of relying on latest URLs. Not only could the URLs change but the content could diverge.

Reviewer: Ideally I would like it to be based on `{graalvm-version}`. If the link breaks due to a bump to the graalvm-version then the guide would also be broken any way.

Is there any tool that can check for broken asciidoc links in the CI? If so, it might be worth enabling it in a different PR

Author: > Ideally I would like it to be based on {graalvm-version}. If the link breaks due to a bump to the graalvm-version then the guide would also be broken any way.

True, it would but then you can fix it for that particular Quarkus / GraalVM combo. IOW, it's something that falls within the graalvm version upgrade. I think things get a lot messier with a version-less URL.

Reviewer: > I think things get a lot messier with a version-less URL.

+1

Author: I've gone for a very simple solution that defines a major.minor version, and use that as part of the full major.minor.micro version, as well as with the existing GraalVM documentation URLs. I also fixed a 22.1 URL that has changed for 22.2.

Reviewer: I am afraid this will result in micro never being bumped.

Grepping the docs I see that we are using the major.minor.micro only to indicate which GraalVM version to use:

```
docs/src/main/asciidoc/gradle-tooling.adoc
388:Make sure to have `GRAALVM_HOME` configured and pointing to GraalVM version {graalvm-version} (Make sure to use a Java 11 version of GraalVM).

docs/src/main/asciidoc/maven-tooling.adoc
377:Make sure to have `GRAALVM_HOME` configured and pointing to GraalVM version {graalvm-version} (Make sure to use a Java 11 version of GraalVM).

docs/src/main/asciidoc/doc-reference.adoc
326:|\{graalvm-version}|{graalvm-version}| Recommended GraalVM version to use.

docs/src/main/asciidoc/building-native-image.adoc
92:Version {graalvm-version} is required. Using the community edition is enough.
```

I believe we could get away by just saying "use the latest major.minor release" instead.

We also use the major.minor.micro in graalvm-flavor (which is the builder-image tag) but especially for this we should be using only major.minor since we want users to use the floating tag.

I will prepare a separate PR for these changes to get some feedback.

Reviewer: Done in https://github.com/quarkusio/quarkus/pull/28843

Author: Makes sense @zakkak. Thanks for adding that PR. I'll rebase this based on that.

The implementation in GraalVM however is different from the HotSpot one,
and there can be significant differences in the runtime behavior.

One of the key differences between HotSpot’s Serial GC and GraalVM’s Serial GC is the way they perform full GC cycles.
In HotSpot the algorithm used is mark-sweep-compact whereas in GraalVM it is mark-copy.
Both need to traverse all live objects,
but in mark-copy this traversal is also used to copy live objects to a secondary space or semi-space.
As objects are copied from one semi-space to another they’re also compacted.
In mark-sweep-compact, the compacting requires a second pass on the live objects.
This makes full GCs in mark-copy more time efficient (in terms of time spent in each GC cycle) than mark-sweep-compact.
The tradeoff mark-copy makes in order to make individual full GC cycles shorter is space.
The use of semi-spaces means that for an application to maintain the same GC performance that mark-sweep achieves (in terms of allocated MB per second),
it requires double the amount of memory.

===== GC Collection Policy

> ===== GC Collection Policy

Reviewer 2: Is the title intentionally duplicated?
